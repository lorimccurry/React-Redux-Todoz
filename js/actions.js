export const ADD_ITEM = 'ADD_ITEM'
export const SET_ITEM_INPUT = 'SET_ITEM_INPUT'
export const SET_ITEMS = 'SET_ITEMS'
