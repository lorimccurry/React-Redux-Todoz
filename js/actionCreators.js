import { SET_ITEM_INPUT, SET_ITEMS } from './actions'

export function setItemInput (itemInput) {
  return { type: SET_ITEM_INPUT, itemInput }
}

export function setItems (items) {
  return { type: SET_ITEMS, items }
}
