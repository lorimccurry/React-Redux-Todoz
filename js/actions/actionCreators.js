import { ADD_ITEM, DELETE_ITEM, SET_ACTIVE_FILTER, SET_EDIT_ID, SET_TOGGLE_COMPLETE, SET_UPDATE_TEXT } from './actions'
import shortid from 'shortid'
import { all } from '../constants'

export function addItem (text) {
  return {
    type: ADD_ITEM,
    payload: {
      item: {
        completed: false,
        id: shortid.generate(),
        text
      }
    }
  }
}

export function deleteItem (deleteID) {
  return {
    type: DELETE_ITEM,
    payload: {
      deleteID
    }
  }
}

export function setActiveFilter (filter = all) {
  return {
    type: SET_ACTIVE_FILTER,
    payload: {
      filter
    }
  }
}

export function setEditID (editID) {
  return {
    type: SET_EDIT_ID,
    payload: {
      editID
    }
  }
}

export function setToggleComplete (itemID) {
  return {
    type: SET_TOGGLE_COMPLETE,
    payload: {
      itemID
    }
  }
}

export function setUpdateText (text) {
  return {
    type: SET_UPDATE_TEXT,
    payload: {
      text
    }
  }
}
