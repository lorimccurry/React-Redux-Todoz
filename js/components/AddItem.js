import React from 'react'
import { connect } from 'react-redux'
import { addItem } from '../actions/actionCreators'
import { FormGroup, FormControl, ControlLabel } from 'react-bootstrap'

class AddItem extends React.Component {
  constructor () {
    super()

    this.state = {
      itemInput: ''
    }

    this.handleChange = this.handleChange.bind(this)
    this.handleSubmit = this.handleSubmit.bind(this)
  }

  handleChange (e) {
    this.setState({itemInput: e.target.value})
  }

  handleSubmit (e) {
    e.preventDefault()
    this.props.dispatchAddItem(this.state.itemInput)
    this.setState({itemInput: ''})
  }

  render () {
    return (
      <form onSubmit={this.handleSubmit}>
        <FormGroup>
          <ControlLabel>Gimme tha todoz:</ControlLabel>
          <FormControl
            type='text'
            placeholder='Get coffee'
            value={this.state.itemInput}
            onChange={this.handleChange}
            autoFocus
            bsSize='lg'
          />
        </FormGroup>
      </form>
    )
  }
}

const { func, string } = React.PropTypes
AddItem.propTypes = {
  dispatchAddItem: func,
  itemInput: string
}

const mapDispatchToProps = (dispatch) => {
  return {
    dispatchAddItem: (text) => {
      dispatch(addItem(text))
    }
  }
}

export const Unwrapped = AddItem

export default connect(null, mapDispatchToProps)(AddItem)
