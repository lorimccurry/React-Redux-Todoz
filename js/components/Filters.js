import React from 'react'
import { connect } from 'react-redux'
import { setActiveFilter } from '../actions/actionCreators'

class Filters extends React.Component {
  constructor () {
    super()

    this.handleClick = this.handleClick.bind(this)
  }

  handleClick (filter) {
    this.props.dispatchActiveFilter(filter)
  }

  render () {
    return (
      <div>
        <button
          className='filter-all'
          onClick={() => this.handleClick('all')}
        >
          All
        </button>
        <button
          className='filter-completed'
          onClick={() => this.handleClick('completed')}
        >
          Completed
        </button>
        <button
          className='filter-uncompleted'
          onClick={() => this.handleClick('uncompleted')}
        >
          Uncompleted
        </button>
      </div>
    )
  }
}

const { func } = React.PropTypes
Filters.propTypes = {
  dispatchActiveFilter: func
}

const mapDispatchToProps = (dispatch) => {
  return {
    dispatchActiveFilter: (filter) => {
      dispatch(setActiveFilter(filter))
    }
  }
}

export const Unwrapped = Filters

export default connect(null, mapDispatchToProps)(Filters)
