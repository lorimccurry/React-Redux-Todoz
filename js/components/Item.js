import React from 'react'
import { connect } from 'react-redux'
import { deleteItem, setEditID, setToggleComplete } from '../actions/actionCreators'
import ImmutablePropTypes from 'react-immutable-proptypes'
import { Button, ButtonToolbar, Checkbox, Row, Col } from 'react-bootstrap'

class Item extends React.Component {
  constructor () {
    super()

    this.handleToggleComplete = this.handleToggleComplete.bind(this)
  }

  handleToggleComplete (e, id) {
    if (e.keyCode === 13 || e.type === 'change') {
      this.props.dispatchToggleComplete(id)
    }
  }

  render () {
    const { item, dispatchDeleteItem, dispatchSetEditID } = this.props
    return (
      <li className='list-group-item'>
        <Row className='show-grid'>
          <Col sm={8} xs={6}>
            <Checkbox
              inline
              type='checkbox'
              checked={item.get('completed')}
              onChange={(e) => this.handleToggleComplete(e, item.get('id'))}
              onKeyDown={(e) => this.handleToggleComplete(e, item.get('id'))}
              name='completed'
            >
              <p className='item-text'>{item.get('text')}</p>
            </Checkbox>
          </Col>
          <Col sm={4} xs={6}>
            <ButtonToolbar>
              <Button
                onClick={() => dispatchSetEditID(item.get('id'))}
                className='edit-btn'
                bsStyle='primary'
                bsSize='lg'
              >
                Edit
              </Button>
              <Button
                onClick={() => dispatchDeleteItem(item.get('id'))}
                className='delete-btn'
                bsStyle='primary'
                bsSize='lg'
              >
                Delete
              </Button>
            </ButtonToolbar>
          </Col>
        </Row>
      </li>
    )
  }
}

const { func } = React.PropTypes
Item.propTypes = {
  item: ImmutablePropTypes.mapContains({
    id: React.PropTypes.string.isRequired,
    text: React.PropTypes.string.isRequired,
    completed: React.PropTypes.bool.isRequired
  }),
  dispatchDeleteItem: func,
  dispatchSetEditID: func,
  dispatchToggleComplete: func
}

const mapDispatchToProps = (dispatch) => {
  return {
    dispatchDeleteItem: (id) => {
      dispatch(deleteItem(id))
    },
    dispatchSetEditID: (id) => {
      dispatch(setEditID(id))
    },
    dispatchToggleComplete: (id) => {
      dispatch(setToggleComplete(id))
    }
  }
}

export const Unwrapped = Item

export default connect(null, mapDispatchToProps)(Item)
