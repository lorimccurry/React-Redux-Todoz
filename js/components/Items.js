import React from 'react'
import Item from './Item'
import EditItem from './EditItem'
import ImmutablePropTypes from 'react-immutable-proptypes'

function Items (props) {
  const { items, editID } = props
  if (!items) {
    return null
  }
  function renderListItems (items, editID) {
    return items.map((item) => {
      if (editID === item.get('id')) {
        return <EditItem
          key={item.get('id')}
          item={item}
        />
      } else {
        return <Item
          key={item.get('id')}
          item={item}
        />
      }
    })
  }

  return (
    <ul className='list-group'>
      {renderListItems(items, editID)}
    </ul>
  )
}

const { string } = React.PropTypes
Items.propTypes = {
  items: ImmutablePropTypes.list,
  editID: string
}

export default Items
