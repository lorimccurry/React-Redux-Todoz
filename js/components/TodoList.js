import React from 'react'
import AddItem from './AddItem'
import Items from './Items'
import Filters from './Filters'
import { connect } from 'react-redux'
import ImmutablePropTypes from 'react-immutable-proptypes'
import { completed } from '../constants'

class TodoList extends React.Component {
  constructor () {
    super()

    this.filteredItems = this.filteredItems.bind(this)
    this.renderListDisplay = this.renderListDisplay.bind(this)
  }

  filteredItems (items, activeFilter) {
    var filteredItems = {
      'completed': () => {
        return items.groupBy(item => item.get(completed)).get(true)
      },
      'uncompleted': () => {
        return items.groupBy(item => item.get(completed)).get(false)
      },
      'default': () => {
        return items
      }
    }
    return (filteredItems[activeFilter] || filteredItems['default'])()
  }

  renderListDisplay (items, editID, activeFilter) {
    const filteredItems = this.filteredItems(items, activeFilter)
    var hasItems = items.size > 0 ? 'yes' : 'default'

    var listRender = {
      'yes': () => {
        return <Items
          items={filteredItems}
          editID={editID}
          />
      },
      'default': () => {
        return <p className='add-msg'>Add some todoz!</p>
      }
    }
    return (listRender[hasItems] || listRender['default'])()
  }

  render () {
    const { items, editID, activeFilter } = this.props
    return (
      <div>
        <AddItem />
        {this.renderListDisplay(items, editID, activeFilter)}
        <Filters />
      </div>
    )
  }
}

const { string } = React.PropTypes
TodoList.propTypes = {
  items: ImmutablePropTypes.list,
  editID: string,
  activeFilter: string
}

const mapStateToProps = (state) => {
  return {
    items: state.todos.items,
    editID: state.todos.editID,
    activeFilter: state.activeFilter.filter
  }
}

export const Unwrapped = TodoList

export default connect(mapStateToProps)(TodoList)
