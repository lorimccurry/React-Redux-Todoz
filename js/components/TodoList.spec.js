import React from 'react'
import { Provider } from 'react-redux'
import { shallow, mount } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import store from '../store/store'
import { addItem, deleteItem, setActiveFilter, setToggleComplete } from '../actions/actionCreators'
import TodoList from './TodoList'
import { Unwrapped as UnwrappedTodoList } from './TodoList'
import Items from './Items'
import Item from './Item'
import EditItem from './EditItem'
import AddItem from './AddItem'
import Filters from './Filters'
import { completed, uncompleted, all } from '../constants'

describe('TodoList', () => {
  it('TodoList snapshot test', () => {
    const stateItems = store.getState().todos.get('items')
    const stateEditID = store.getState().todos.get('editID') 
    const component = shallow(<UnwrappedTodoList items={stateItems} editID={stateEditID} />)
    const tree = shallowToJson(component)
    expect(tree).toMatchSnapshot()
  })

  it('renders AddItem component', () => {
    const stateItems = store.getState().todos.get('items')
    const stateEditID = store.getState().todos.get('editID') 
    const component = shallow(<UnwrappedTodoList items={stateItems} editID={stateEditID} />)
    expect(component.find(AddItem).length).toEqual(1)
  })

  it('should start with an empty list', () => {
    const stateItems = store.getState().todos.get('items')
    const stateEditID = store.getState().todos.get('editID') 
    const component = shallow(<UnwrappedTodoList items={stateItems} editID={stateEditID} />)
    const p = component.find('p')
    expect(component.find('li').length).toEqual(0)
    expect(p.length).toEqual(1)
    expect(p.text()).toEqual('Add some todoz!')
    expect(stateItems.size).toEqual(0)
  })

  it('renders instruction message when no items', () => {
    const stateItems = store.getState().todos.get('items')
    const stateEditID = store.getState().todos.get('editID') 
    const component = shallow(<UnwrappedTodoList items={stateItems} editID={stateEditID} />)
    expect(stateItems.size).toEqual(0)
    expect(component.contains(<p className='add-msg'>Add some todoz!</p>)).toEqual(true)
  })

  it('renders Items when items and add message when deleted and no items', () => {
    const component = mount(<Provider store={store}><TodoList /></Provider>)
    store.dispatch(addItem('get beer'))

    expect(component.containsMatchingElement(<Items />)).toEqual(true)
    expect(component.contains(<p className='add-msg'>Add some todoz!</p>)).toEqual(false)

    const deleteID = store.getState().todos.get('items').get(0).get('id')
    store.dispatch(deleteItem(deleteID))

    expect(component.containsMatchingElement(<Items />)).toEqual(false)
    expect(component.contains(<p className='add-msg'>Add some todoz!</p>)).toEqual(true)
  })

  it('renders Filters', () => {
    const stateItems = store.getState().todos.get('items')
    const stateEditID = store.getState().todos.get('editID') 
    const component = shallow(<UnwrappedTodoList items={stateItems} editID={stateEditID} />)
    expect(component.containsMatchingElement(<Filters />)).toEqual(true)
  })
})

describe('TodoList integration testing', () => {
  it('adds items to the store and view', () => {
    const component = mount(<Provider store={store}><TodoList /></Provider>)

    expect(component.containsMatchingElement(<Items />)).toEqual(false)
    expect(component.contains(<p className='add-msg'>Add some todoz!</p>)).toEqual(true)
    expect(store.getState().todos.get('items').size).toEqual(0)

    const addInput = component.find('input[type="text"]')
    addInput.simulate('change', {target: {value: 'get beer'}})

    const addForm = component.find('form')
    addForm.simulate('submit')

    expect(component.containsMatchingElement(<Items />)).toEqual(true)
    expect(component.contains(<p>Add some todoz!</p>)).toEqual(false)
    expect(store.getState().todos.get('items').size).toEqual(1)
    expect(component.containsMatchingElement(<Item />)).toEqual(true)
    expect(component.containsMatchingElement(<EditItem />)).toEqual(false)
    expect(component.find(Item).length).toEqual(1)
    expect(addInput.prop('value')).toEqual('')

    addInput.simulate('change', {target: {value: 'get coffee'}})
    addForm.simulate('submit')
    expect(component.find(Item).length).toEqual(2)
  })

  it('deletes items from store and view', () => {
    const component = mount(<Provider store={store}><TodoList /></Provider>)
    expect(component.find(Item).length).toEqual(2)

    let item = component.find(Item).first()
    let  deleteBtn = item.find('.delete-btn')
    deleteBtn.simulate('click')

    expect(component.find(Item).length).toEqual(1)
    expect(store.getState().todos.get('items').size).toEqual(1)

    item = component.find(Item)
    deleteBtn = item.find('.delete-btn')
    deleteBtn.simulate('click')

    expect(component.containsMatchingElement(<Items />)).toEqual(false)
    expect(component.contains(<p className='add-msg'>Add some todoz!</p>)).toEqual(true)
    expect(store.getState().todos.get('items').size).toEqual(0)
  })

  it('edits item text and store', () => {
    const component = mount(<Provider store={store}><TodoList /></Provider>)
    store.dispatch(addItem('eat'))
    store.dispatch(addItem('sleep'))
    store.dispatch(addItem('code'))

    let item = component.find(Item).first()
    expect(item.find('.item-text').text()).toEqual('eat')

    const editBtn = item.find('.edit-btn')
    editBtn.simulate('click')

    expect(component.find(EditItem).length).toEqual(1)
    expect(component.find(Item).length).toEqual(2)

    const updateInput = component.find('.update-input')
    expect(updateInput.prop('value')).toEqual('eat')

    const event = {target: {type: 'text', name: 'text', value: 'live'}}
    updateInput.simulate('change', event)

    const itemStoreText = store.getState().todos.get('items').get(0).get('text')
    const itemValue = updateInput.prop('value')
    expect(itemStoreText).toEqual('live')
    expect(itemValue).toEqual('live')

    const updateBtn = component.find('.update-btn')
    updateBtn.simulate('click')

    item = component.find(Item).first()

    expect(item.find('.item-text').text()).toEqual('live')
    expect(component.find(EditItem).length).toEqual(0)
    expect(component.find(Item).length).toEqual(3)
  })

  it('toggles item checkbox state and view', () => {
    const component = mount(<Provider store={store}><TodoList /></Provider>)

    let item = component.find('Item').first()
    const itemCheckbox = item.find('input[name="completed"]')
    const itemCheckboxStore = store.getState().todos.get('items').get(0).get(completed)
    const itemCheckboxChecked = itemCheckbox.prop('checked')
    expect(itemCheckboxStore).toEqual(false)
    expect(itemCheckboxChecked).toEqual(false)

    itemCheckbox.simulate('change')
    let updatedItemCheckboxStore = store.getState().todos.get('items').get(0).get(completed)
    let updatedItemCheckboxChecked = itemCheckbox.prop('checked')

    expect(updatedItemCheckboxStore).toEqual(true)
    expect(updatedItemCheckboxChecked).toEqual(true)

    itemCheckbox.simulate('change')
    updatedItemCheckboxStore = store.getState().todos.get('items').get(0).get(completed)
    updatedItemCheckboxChecked = itemCheckbox.prop('checked')

    expect(updatedItemCheckboxStore).toEqual(false)
    expect(updatedItemCheckboxChecked).toEqual(false)
  })

  it('filters items', () => {
    const component = mount(<Provider store={store}><TodoList /></Provider>)
    // make item1 completed 
    let item1 = store.getState().todos.get('items').get(0)
    store.dispatch(setToggleComplete(item1.get('id')))

    item1 = store.getState().todos.get('items').get(0)
    let item2 = store.getState().todos.get('items').get(1)
    let item3 = store.getState().todos.get('items').get(2)
    let items = store.getState().todos.get('items')

    expect(items.size).toEqual(3)
    expect(item1.get(completed)).toEqual(true)
    expect(item2.get(completed)).toEqual(false)
    expect(item3.get(completed)).toEqual(false)

    // test 3 filters
    store.dispatch(setActiveFilter(uncompleted))
    expect(component.find(Item).length).toEqual(2)

    store.dispatch(setActiveFilter(completed))
    expect(component.find(Item).length).toEqual(1)

    store.dispatch(setActiveFilter(all))
    expect(component.find(Item).length).toEqual(3)

    // make all items uncompleted
    item1 = store.getState().todos.get('items').get(0)
    store.dispatch(setToggleComplete(item1.get('id')))

    item1 = store.getState().todos.get('items').get(0)
    expect(item1.get(completed)).toEqual(false)

    // handle an undefined items or filteredItems case
    store.dispatch(setActiveFilter(completed))
    expect(component.find(Item).length).toEqual(0)
  })
})
