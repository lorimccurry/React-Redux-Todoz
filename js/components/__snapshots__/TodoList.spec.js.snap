// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList TodoList snapshot test 1`] = `
<div>
  <Connect(AddItem) />
  <p
    className="add-msg"
  >
    Add some todoz!
  </p>
  <Connect(Filters) />
</div>
`;
